doc-site/docs/architecture/distributed_sequencer.md

```diff
@@ -0,0 +1,1437 @@
+
+For the current block height, a position on the ring is calculated from a deterministic function.  The node that is closest to that position on the ring is selected as the coordinator.  If that node is not available, then the next closest node is selected. If that node is not available, then the next closest again is selected, and so on.
+
+For each Pente domain instance ( smart contract API) the coordinator is selected by choosing one of the nodes in the configured privacy group of that contract
+
+ - the selector is a pure function where the inputs are the node names + the current block number and the output is the name of one of the nodes
+ - the function will return the same output for a range of `n` consecutive blocks ( where `n` is a configuration parameter of the policy)
+ - for the next range of `n` blocks, the function will return a different ( or possibly the same) output
+ - over a large sample of ranges, the function will have an even distribution of outputs
+ - the function will be implemented by taking a hash of each node name, taking the hash of the output of `b/n` (where b is the block number and `n` is the range size) rounded down to the nearest integer and feeding those into a hashring function.
+
+![image](../images/distributed_sequencer_hashring.svg){ width="500" }
+
+In the example here, lets say we chose a hash function with a numeric output between 0 and 360 where the hash of node A's name happens to be 315, node B is 45, node C is 225 and node D is 135.  Lets say we have a range size of 4, therefore for block numbers 0-3, we can calculate that they belong to range number 0, blocks 4-7 belong to range number 1 and so on.  Lets say that the hash function, when applied to `range0` returns 310.  When applied to `range1` returns 55 and when applied to `range2` returns 120.  Interpreting all of those hash output as degrees and plotting them on the circle ( with 0° at the top). We can then deterministically calculate that nodeA is selected while the block height is between 0 and 3, node B is selected when the block height is between 4 and 7 and node D is selected while the block height is between 8 an 11
+
+
+### Node availability detection
+
+When a node starts acting as the coordinator, then it  periodically sends a heartbeat message to all nodes that for which it has an active delegation and will accept delegation requests from any other node in the group. 
+
+All sender nodes keep track of the current selected coordinator ( either by reevaluating the deterministic function periodically or caching the result until block height exceeds the current range boundary).  If they fail to receive a heartbeat message from the current coordinator, then they will chose the next closest node in the hashring (e.g. by reevaluating the hashring function with the unavailable nodes removed).
```

Comment:
What is the intended behavior of the algorithm in this scenario?

- Highest ranked coordinator is B
- D delegates transaction to B
- B stops for a period
    - Longer than the heartbeat interval
    - Shorter than it takes for the block-range to change
- A is the second-highest scored coordinator
- D delegates transaction to A (because it gives up on B)
- B becomes available again
- C delegates transaction to B

At this point D has delegated a TX to A, but C has delegated to B.
I haven't (yet at this point in the spec) understood what would cause A and B to notice that they are both active (because at the time A became the coordinator, B was not available to ask).

Reply:
Ah, ok found this statement below:

> If Node A comes back online, then it will begin transmitting heartbeat messages again because it has selected itself as coordinator given the current block height.

This seems complex to implement - what would cause A to evaluate that it selected itself as the coordinator? It does not have any transactions after the the restart. Why would this contract (of the infinite potential contracts) be one after restart it would choose to select itself as a coordinator?